signing_method.go

```diff
@@ -14,22 +14,30 @@ type SigningMethod interface {
 	Alg() string                                                   // returns the alg identifier for this method (example: 'HS256')
 }
 
-// RegisterSigningMethod registers the "alg" name and a factory function for signing method.
-// This is typically done during init() in the method's implementation
-func RegisterSigningMethod(alg string, f func() SigningMethod) {
+// Register internally creates a new function that returns the given
+// SigningMethod. The function will be stored as a value in a thread safe map,
+// where the algorithm name is the key.
+func Register(m SigningMethod) {
```

Comment:
Just another small nitpick. Can we have a different name than just `Register`? As we do not have any sub packages, having a `Register` function directly at the public package level might be confusing.

Also, is it intended to have this exported? I guess to support external packages? Or should we direct them to `RegisterSigningMethod` (if this is actually still working now?). In this case, we might decide to make it private.

Reply:
I'd vote to keep `Register` unexported.  Consumers already have access to `RegisterSigningMethod` which does the same thing, albeit in a less friendly way.

But I think having a smaller API surface and a single way is better overall.

Reply:
I would prefer if the `Register(m SigningMethod)` function was named `RegisterSigningMethod(m SigningMethod)`, but `RegisterSigningMethod` is already used for `RegisterSigningMethod(alg string, f func() SigningMethod)`, and renaming that function or making it lowercase would change the API.

I think it would be helpful if `Register(m SigningMethod)` was exported, because it's a nice way to register new `SigningMethod`s with `golang-jwt/jwt`. Used from another package, it would have the context of `jwt`, like this: `jwt.Register(mySigningMethod)`, so it would not be without context.

I will happily change this pull request if there are suggestions for a better name than `Register` (or if it would be possible to change the API so that the exported function would be `RegisterSigningMethod(m SigningMethod)` instead of `RegisterSigningMethod(alg string, f func() SigningMethod)`. I feel like just unexporting `Register` by renaming it to `register` is a bit tame solution, and would not provide the best API for users.

Reply:
I think the name is fine, and what you've proposed absolutely makes sense. 

The issue is adding yet another way to do something. As a consumer I know have to understand the difference between `RegisterSigningMethod` vs `Register` and that's already too much work.

I agree this package has a lot of room for improvement, but if we continue duplicating existing functions / methods then it becomes confusing for both old and new users, the former needs to understand if there is an actual improvement in the newly added API  and new users need to understand the difference between the two.

I think it would make sense to replace `RegisterSigningMethod` with `Register`  in a `/v5` release.

The alternative is to add a `Register` as you proposed with a deprecation notice to `RegisterSigningMethod`.

Reply:
Thanks for the insightful response, @mfridman. I added a deprecation comment to the pull request.

I also believe that replacing `RegisterSigningMethod` with the `Register` function would make sense for a `v5` release.

Reply:
@oxisto and co. Did you have further opinions on `RegisterSigningMethod` and `Register` ? I know we're bideshedding here a bit, but it's worth discussing as it sets a precedent for the package as a whole.

I still worry about the number of PRs introducing new exportable functions. Albeit adding value, they do introduce complexity and duplication into the codebase.

Reply:
Alternatively, the `RegisterSigningMethod` and `Register` method can be combined into a single `RegisterSigningMethod` by using duck typed arguments.

This way the API is kept stable, type checking is still in place, it is possible to just pass in a single struct that implements the SigningMethod interface, and all tests pass here:

```go
func RegisterSigningMethod(algOrSigningMethods ...interface{}) error {
	switch len(algOrSigningMethods) {
	case 1:
		// One argument is given, check if it is a SigningMethod
		m, ok := algOrSigningMethods[0].(SigningMethod)
		if !ok {
			return errors.New("one argument is given, and it does not implement SigningMethod")
		}
		signingMethodLock.Lock()
		defer signingMethodLock.Unlock()
		signingMethods[m.Alg()] = func() SigningMethod { return m }
		return nil
	case 2:
		// Two arguments are given, check if it is a string and a func() SigningMethod
		alg, ok := algOrSigningMethods[0].(string)
		if !ok {
			return errors.New("two arguments are given and the first argument is not a string")
		}
		f, ok := algOrSigningMethods[1].(func() SigningMethod)
		if !ok {
			return errors.New("two arguments are given and the second argument is not a function that returns a SigningMethod")
		}
		signingMethodLock.Lock()
		defer signingMethodLock.Unlock()
		signingMethods[alg] = f
		return nil
	}
	return errors.New("too few or too many arguments are given")
}
```

Reply:
@mfriedman:

@oxisto and others does not appear to have further opinions on this